### PR TITLE
[FLINK-4190] Generalise RollingSink to work with arbitrary buckets

### PR DESCRIPTION
**Pull request by joshfg**
I've created a new bucketing package with a BucketingSink, which improves on the existing RollingSink by enabling arbitrary bucketing, rather than just rolling files based on system time.

The main changes to support this are:

- The Bucketer interface now takes the sink's input element as a generic parameter, enabling us to bucket based on attributes of the sink's input.
- While maintaining the same rolling mechanics of the existing implementation (e.g. rolling when the file size reaches a threshold), the sink implementation can now have many 'active' buckets at any point in time. The checkpointing mechanics have been extended to support maintaining the state of multiple active buckets and files, instead of just one.
- For use cases where the buckets being written to are changing over time, the sink now needs to determine when a bucket has become 'inactive', in order to flush and close the file. In the existing implementation, this is simply when the bucket path changes. Instead, we now determine a bucket as inactive if it hasn't been written to recently. To support this there are two additional user configurable settings: inactiveBucketCheckInterval and inactiveBucketThreshold.

### CODE REVIEW

**Comment by aljoscha** [comment]
Very good work!

I know we discussed before whether to check for inactivity in a different thread or in `invoke()`. There's actually a third option that I'm showcasing in the PR I did against your PR. 😃 The StreamTask already has a TimerService that can be set for testing. If we use the appropriate methods in the bucketing sink then we get testability with a settable clock for free.

I also added a `BucketSinkTest`. It think it would be good if the `BucketSinkITCase` and `BucketingSinkMultipleActiveBucketsCase` could be folded into this one because having the ITCases means having a lot of overhead and our build is already taking quite long.

What do you think?

**Comment by joshfg** [comment]
Thanks! Oh nice, this looks like a better solution for checking for bucket inactivity...
For the tests, is there any reason not to fold all of those tests into the new `BucketingSinkTest`? Currently there's 4: (BucketingSinkITCase, BucketingSinkFaultToleranceITCase, BucketingSinkFaultTolerance2ITCase, BucketingSinkMultipleActiveBucketsCase)

Also, do you know what's the purpose of using MiniDFSCluster in the tests? Could we rewrite the other tests in the same way as your example test, without running a local HDFS cluster?

**Comment by aljoscha** [comment]
Not really, the reason for having an ITCase is just that they really exercise the sink embedded in a proper Flink job, which might bring up interactions that where overlooked when writing a test case. I think with proper test cases we could go completely without an ITCase. Which also improves testing time.

The purpose of the `MiniDFSCluster` is to test the sink against an actual HDFS cluster because there is some stuff in there that would only be tested when using HDFS. For example, the truncate support.

tl;dr Go for it and move everything to a test case. 😃

**Comment by joshfg** [comment]
Ah I see, that makes sense.
I've began refactoring the tests here: https://github.com/joshfg/flink/blob/flink-4190/flink-streaming-connectors/flink-connector-filesystem/src/test/java/org/apache/flink/streaming/connectors/fs/bucketing/BucketingSinkTest.java

But have run into this strange exception in the `testNonRollingSequenceFileWithoutCompressionWriter` test:
```
java.lang.IllegalStateException: Key Class has not been initialized.
	at org.apache.flink.streaming.connectors.fs.SequenceFileWriter.open(SequenceFileWriter.java:84)
	at org.apache.flink.streaming.connectors.fs.bucketing.BucketingSink.openNewPartFile(BucketingSink.java:500)
	at org.apache.flink.streaming.connectors.fs.bucketing.BucketingSink.invoke(BucketingSink.java:396)
	at org.apache.flink.streaming.api.operators.StreamSink.processElement(StreamSink.java:39)
	at org.apache.flink.streaming.util.OneInputStreamOperatorTestHarness.processElement(OneInputStreamOperatorTestHarness.java:226)
	at org.apache.flink.streaming.connectors.fs.bucketing.BucketingSinkTest.testNonRollingSequenceFileWithoutCompressionWriter(BucketingSinkTest.java:220)
```

Any ideas what would cause that? I've copied the HDFS cluster initialisation exactly as it was in the original tests...

**Comment by aljoscha** [comment]
This is caused by `setInputType()` not being called on `SequenceFileWriter`. In the test, you can call `setInputType` on the `BucketingSink` once with the input `TypeInformation` and a `new ExecutionConfig()`. That should do the trick.

**Comment by joshfg** [comment]
That works, thanks! :)

**Comment by joshfg** [comment]
Ok I've migrated `BucketingSinkITCase` and `BucketingSinkMultipleActiveBucketsCase` over to `BucketingSinkTest` using the test harness with `TimeServiceProvider`. I've left the two fault tolerance IT cases as they are because it looks like they need to run a proper Flink job with a custom source/mappers. Does that sound OK?

If you think it's ready to merge, should I move all the commits into a single commit for FLINK-4190?

**Comment by aljoscha** [comment]
Thanks for moving the tests! It should be ok to leave these other two ITCases.

I'll merge the commits into one once I can find some time to look at the tests and merge it.

**Comment by joshfg** [comment]
Hi Aljoscha, just wanted to remind you about this - any idea when the changes will be merged in? Thanks!

**Comment by aljoscha** [comment]
Hi,
I'm very for the delays! I still have this sitting at the top of my list and I'm hoping to get this in by beginning of next week.

**Comment by aljoscha** [comment]
I merged it, could you please close this PR if it's not closed automatically.

Thanks for your work, and sorry again for taking so long to get it in! 👍

**Comment by joshfg** [comment]
That's great, thanks Aljoscha!
